src/Nethermind/Nethermind.Evm/TransactionSubstate.cs

```diff
@@ -67,41 +68,47 @@ public TransactionSubstate(EvmExceptionType exceptionType, bool isTracerConnecte
             return;
 
         ReadOnlySpan<byte> span = Output.Span;
-        Error = TryGetErrorMessage(span)
-                ?? DefaultErrorMessage(span);
+        Error = TryGetErrorMessage(span) ?? DefaultErrorMessage(span);
     }
 
-    private string DefaultErrorMessage(ReadOnlySpan<byte> span)
-    {
-        return string.Concat(RevertedErrorMessagePrefix, span.ToHexString(true));
-    }
+    private string DefaultErrorMessage(ReadOnlySpan<byte> span) => string.Concat(RevertedErrorMessagePrefix, span.ToHexString(true));
 
     private unsafe string? TryGetErrorMessage(ReadOnlySpan<byte> span)
     {
-        if (span.Length < RevertPrefix + sizeof(UInt256) * 2)
+        const int fieldLength = EvmPooledMemory.WordSize;
+        if (span.Length < RevertPrefix + fieldLength * 2)
         {
             return null;
         }
 
         try
         {
-            int start = (int)new UInt256(span.Slice(RevertPrefix, sizeof(UInt256)), isBigEndian: true);
-            if (checked(RevertPrefix + start + sizeof(UInt256)) > span.Length)
-            {
-                return null;
-            }
+            int start = (int)new UInt256(span.Slice(RevertPrefix, fieldLength), isBigEndian: true);
 
-            int length = (int)new UInt256(span.Slice(RevertPrefix + start, sizeof(UInt256)), isBigEndian: true);
-            if (checked(RevertPrefix + start + sizeof(UInt256) + length) != span.Length)
+            int lengthStart = RevertPrefix + start;
+            int lengthEnd = checked(lengthStart + fieldLength);
+            if (lengthEnd <= span.Length)
             {
-                return null;
+                int messageLength = (int)new UInt256(span.Slice(lengthStart, fieldLength), isBigEndian: true);
+                int messageEnd = checked(lengthEnd + messageLength);
+                if (messageEnd <= span.Length)
+                {
+                    ReadOnlySpan<byte> message = span.Slice(lengthEnd, messageLength);
+
+                    if (messageEnd == span.Length)
+                    {
+                        return string.Concat(RevertedErrorMessagePrefix, message.ToHexString(true));
```

Comment:
we are returning hex string here, but UTF8 later - inconsistent!

Reply:
On Geth code (https://github.com/ethereum/go-ethereum/blob/abe3fca1deb0eca9245cccef3a9c637c57b79f7e/accounts/abi/abi.go#L275C1-L311) there are two cases:
- Call to `Error(string)`, decode value as UTF8 (current PR, available in Erigon).
- Call to `Panic(uint256)`, use the value as a code for a panic reasons map (current PR, not available in Erigon).

If neither is the case, then it does not provide an error message. Still, both Erigon and Geth will always include the data as hex string. For Nethermind then I would not change the default to use UTF8 everywhere since it might be the wrong thing to do.